### PR TITLE
flite: add audiobackend option

### PR DESCRIPTION
**Pull request by getchoo**
###### Description of changes

this adds an `audioBackend` option to flite - letting users specify what audio backend they wish to use for building - along with making the package build against pulseaudio by default instead of alsa. this follows the example of other distros such as [fedora](https://src.fedoraproject.org/rpms/flite/blob/32cd5587c1b8744be0f0fe932048fc97b5e9e914/f/flite.spec#_58) and allows flite to work in minecraft (see https://github.com/NixOS/nixpkgs/pull/238645) and other software where flite built against pulse is preferred

i also noticed this package doesn't have a maintainer, so i added myself :)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/flite/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch }:
+{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch, libpulseaudio, withPulse ? false }:
```
Comment (Scrumplex) [inline pkgs/development/libraries/flite/default.nix:1]: I think this should be more generic for longevity. I could see a PipeWire-based backend in the future. Or someone might need a package with OSS support as well in the future.

So maybe just make this a string like 
```
audioBackend ? "alsa"
```

And add the corresponding dependencies depending on the value.

**Comment by Scrumplex** [comment]
Backporting to 23.05, as I need it for Prism Launcher.

File: pkgs/development/libraries/flite/default.nix (hunk 1 above)
Comment (getchoo) [reply pkgs/development/libraries/flite/default.nix]: awesome idea! i just implemented this :)

File: pkgs/development/libraries/flite/default.nix [hunk 2]
```diff
@@ -1,4 +1,6 @@
-{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch }:
+{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch, libpulseaudio, audioBackend ? "alsa" }:
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/flite/default.nix:1]:
```suggestion
{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch, libpulseaudio, audioBackend ? "pulseaudio" }:
```

**Review by Scrumplex — APPROVED** [review]

File: pkgs/development/libraries/flite/default.nix [hunk 3]
```diff
@@ -11,7 +13,10 @@ stdenv.mkDerivation rec {
     sha256 = "1n0p81jzndzc1rzgm66kw9ls189ricy5v1ps11y0p2fk1p56kbjf";
   };
 
-  buildInputs = lib.optionals stdenv.isLinux [ alsa-lib ];
+  buildInputs = let
+    audio = lib.optionals (audioBackend == "alsa") [ alsa-lib ]
+    ++ lib.optionals (audioBackend == "pulseaudio") [ libpulseaudio ];
+  in lib.optionals stdenv.isLinux audio;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/flite/default.nix:19]:
```suggestion
  buildInputs = lib.optional (stdenv.isLinux && audioBackend == "alsa") alsa-lib
    ++ lib.optional (stdenv.isLinux && audioBackend == "pulseaudio") libpulseaudio;
```

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -20598,6 +20598,8 @@ with pkgs;
 
   flite = callPackage ../development/libraries/flite { };
 
+  flite-pulse = callPackage ../development/libraries/flite { audioBackend = "pulseaudio"; };
+
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20602]:
```suggestion
```

**Review by SuperSandro2000 — COMMENTED** [review]
Why should we prefer alsa over pulse? pulse has more features than alsa. Also I don't think we need an extra package unless it is used somewhere.

**Comment by getchoo** [comment]
> Why should we prefer alsa over pulse?

that's what i thought too but figured i should keep it since it was already the default

> Also I don't think we need an extra package unless it is used somewhere

agreed, especially with pulse as the default

**Review by Scrumplex — APPROVED** [review]
Not sure if changing the backend could break any downstream packages. But changes look okay to me.

**Comment by Scrumplex** [comment]
(This needs to go to staging though)

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Why would this rebuild stdenv :thinking: 

@ofborg eval

**Comment by claike** [comment]
Any progress on this PR?

**Comment by Scrumplex** [comment]
> Any progress on this PR?

Still waiting for reviews

File: pkgs/development/libraries/flite/default.nix [hunk 5]
```diff
@@ -1,4 +1,7 @@
-{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch }:
+{ lib, stdenv, fetchFromGitHub, alsa-lib, fetchpatch, libpulseaudio, audioBackend ?
+"pulseaudio" }:
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/flite/default.nix:2]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, alsa-lib
, fetchpatch
, libpulseaudio
, audioBackend ? "pulseaudio"
}:
```

**Pull request merged**